### DISCUSSION TITLE
"Future already completed" when using AsyncValue.guard in AsyncNotifier

### DISCUSSION THREAD

**Discussion by hibix43** [rrousselGit/riverpod#2502] [Q&A]
"Future already completed" occurs when AsyncValue.guard is used in AsyncNotifier.
Below is a sample and the error that occurred. Is there a solution to this problem?

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'main.g.dart';

void main() async {
  runApp(
    const ProviderScope(
      child: MaterialApp(
        home: App(),
      ),
    ),
  );
}

class App extends ConsumerWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          onPressed: () async {
            await ref.read(sampleNotifierProvider.notifier).fetch();
          },
          child: const Text('Fetch'),
        ),
      ),
    );
  }
}

class FakeDatabase {
  Future<void> fetch() async =>
      Future.delayed(const Duration(seconds: 1), () {});
}

@riverpod
Future<FakeDatabase> fakeDatabase(FakeDatabaseRef ref) async => FakeDatabase();

@riverpod
class SampleNotifier extends _$SampleNotifier {
  @override
  FutureOr<void> build() {}

  Future<void> fetch() async {
    state = const AsyncValue.loading();
    state = await AsyncValue.guard(() async {
      final database = await ref.read(fakeDatabaseProvider.future);
      await database.fetch();
    });
  }
}
``` 

```
E/flutter (12958): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Bad state: Future already completed
E/flutter (12958): #0      _AsyncCompleter.complete (dart:async/future_impl.dart:35:31)
E/flutter (12958): #1      FutureHandlerProviderElementMixin.onData
base.dart:271
E/flutter (12958): #2      AsyncData.map
common.dart:326
E/flutter (12958): #3      FutureHandlerProviderElementMixin.state=
base.dart:204
E/flutter (12958): #4      AsyncNotifierBase.state=
async_notifier.dart:54
E/flutter (12958): #5      SampleNotifier.fetch
main.dart:50
E/flutter (12958): <asynchronous suspension>
E/flutter (12958): #6      App.build.<anonymous closure>
main.dart:26
E/flutter (12958): <asynchronous suspension>
E/flutter (12958):
```

**Comment by AhmedLSayed9** [comment]
You're expected to return `Future<FakeDatabase>` but you're returning an instance of `FakeDatabase` class that is not a future, instead do:
```dart
@riverpod
FakeDatabase fakeDatabase(FakeDatabaseRef ref) => FakeDatabase();
```

**Reply by hibix43** [reply]
@AhmedLSayed9 
Thank you for your response.
Sorry, the sample was incorrect.
In the actual code, we need to load the JSON in the assets.
(e.g. `await rootBundle.loadString(Assets.jsons.databaseConfigJson);`)
 This is why I have to use `Future`. I didn't give you enough information, sorry.

**Comment by AhmedLSayed9** [comment]
@hibix43 Despite using `Future<FakeDatabase>` isn't proper but it's not the reason for your issue anyway.

It's happening because you're updating state after the provider is disposed.
_Note:_ The error description has been updated in latest versions.

You can:
1. temporary listen for the provider until the future is complete:
```dart
        child: ElevatedButton(
          onPressed: () async {
            final sub =
                ref.listenManual(sampleNotifierProvider.notifier, (_, __) {});
            await sub.read().fetch();
            sub.close();
          },
          child: const Text('Fetch'),
        ),
```
2.  Check if the provider is mounted before setting state after the future call: [check here](https://github.com/rrousselGit/riverpod/issues/1879#issuecomment-1303189191)

**Reply by hibix43** [reply]
Thank you for sharing the cause and workaround. I will use that workaround until the cause is resolved.

**Reply by MuKhAlm** [reply]
I have used approach 1. But it seems that my [sampleNotifierProvider] is still being disposed before the future is complete...

**Comment by erik-wayhaven** [comment]
I am running into this as well. This seems like a common usecase. Are there plans to fix it?

I created this extension method for now:
```dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

extension SafeWidgetRef on WidgetRef {
  Future<Tr> safeRead<Tn, Tr>(Refreshable<Tn> notifier, Future<Tr> Function(Tn) call) async {
    final sub = listenManual(notifier, (_, __) {});
    final res = await call(sub.read());
    sub.close();
    return res;
  }
}
```

**Comment by nazrinharris** [comment]
Unsure if this is related, but I've also ran into the same issue. It happened with this code:
```dart
@riverpod
class LocalLoginWithEmailAndPassword extends _$LocalLoginWithEmailAndPassword {
  @override
  FutureOr<Option<PecuniaUser>> build() {
    return const Option.none();
  }

  Future<void> localLoginWithEmailAndPassword({
    required String email,
    required String password,
  }) async {
    state = const AsyncValue.loading();

    final failureOrPecuniaUserAndSession = await ref
        .read(authRepoProvider)
        .localLoginWithEmailAndPassword(
          email: email,
          password: password,
        )
        .run();

    return failureOrPecuniaUserAndSession.fold(
      (failure) => state = AsyncValue.error(failure, failure.stackTrace),
      (r) => state = AsyncValue.data(Option.of(r.user)),
    );
  }
}
```
Which caused a `Bad State: Future already completed` error in `(r) => state = AsyncValue.data(Option.of(r.user)),`. The change that fixes this for me is the build method:
```dart
@override
  Future<Option<PecuniaUser>> build() async {
    return const Option.none();
  }
```
Seems like the `async` is needed in the `build` method?

**Reply by maroafenogho** [reply]
Weird how this worked.
The error disappeared without a trace.

**Reply by mfrischbutter** [reply]
Omg, had the same issue, and debugged this issue for almost an hour. Adding `async` to the `build()` method worked!
This is a bug and needs to be addressed in issues.

**Reply by mfrischbutter** [reply]
You should also mark nazrinharris answer green and not the other one. It was me irritating several times, as I read this discussion.

**Reply by maroafenogho** [reply]
I think only the person who asked the original question can do that.
@mfrischbutter

**Reply by hibix43** [reply]
Thanks. I changed the answer mark.

**Reply by AhmedLSayed9** [reply]
This was a false negative. Provider should throw when setting state after it's being disposed.
It has been fixed at riverpod 3 "dev" and the solution should no longer works.

**Reply by AhmedLSayed9** [reply]
You need to either listen to the provider while the method is running "at my answer above", or check if the provider is mounted before setting state after the future call. I've updated my answer to include both solutions.

**Reply by agordeev** [reply]
@hibix43 please unmark this as accepted answer. This workaround doesn't work anymore.

**Reply by hibix43** [reply]
I have unmarked. Thank you for letting me know that this workaround no longer works, @agordeev.

**Comment by alainjr10** [comment]
Well, this is weird, i had this issue too because my provider was being disposed immediately the button was clicked before any async function was called (you can check this by simply logging some results in the onDispose function of the build method). 

`  @override
  FutureOr<List<Product>> build() async {
    ref.onDispose(() {
     log( 'provider disposed');
    });
    return [];
  }
`

I got to understand that, since the providers are autoDispose by default, since you aren't explicitly listening/watching the provider anywhere in your widget, the provider is inactive and becomes disposed immediately the button is clicked. So I fixed this by keeping the provider active by at least watching it's state in the build method like so

`  final searchedProds = ref.watch(searchProductsNotifierProvider);`

you could then use the searchProds AsyncValue to build your widget based on its state... or not

**Comment by mokamhawy** [comment]
The solution for me was to add `async`
The bad code
```dart 
  @override
  FutureOr<void> build() {}
```

The good code (issue fixed)
```dart 
  @override
  FutureOr<void> build() async {}
```

**Reply by JakobProssinger** [reply]
Same issue here. But I also had to add a "return null" to the build function above.

**Comment by VDR-BeemEnergy** [comment]
Hello,

I’m currently working on a project using Riverpod and a lot of AutoDipose. 
When adding a Crahslytics reporting, I came across two issues occurring a lot.
 
<details><summary>FutureHandlerProviderElementMixin.onError</summary>

```
Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: Bad state: Future already completed
       at _Completer.completeError(dart:async)
       at FutureHandlerProviderElementMixin.onError(base.dart:260)
       at AsyncError.map(common.dart:524)
       at FutureHandlerProviderElementMixin.state=(base.dart:212)
       at AsyncNotifierBase.state=(async_notifier.dart:66)
```

</details>

and
<details><summary>FutureHandlerProviderElementMixin.onData</summary>

```
Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: Bad state: Future already completed
       at _AsyncCompleter.complete(dart:async)
       at FutureHandlerProviderElementMixin.onData(base.dart:289)
       at AsyncData.map(common.dart:346)
       at FutureHandlerProviderElementMixin.state=(base.dart:212)
       at AsyncNotifierBase.state=(async_notifier.dart:66)
```
</details>

The two seem related to AutoDisposeAsyncNotifierProvider being created and disposed very quickly, where state is assigned to a closed instance.

But while trying to understand, I came to the file riverpod-2.6.1/lib/src/async_notifier/base.dart:

<details><summary>  void onError(AsyncError<T> value, {bool seamless = false})</summary>

```
final completer = _futureCompleter;
if (completer != null && completer.isCompleted) {
  completer
    // TODO test ignore
    ..future.ignore()
    ..completeError(
      value.error,
      value.stackTrace,
    );
  _futureCompleter = null;
  // TODO SynchronousFuture.error
}
```
</details>

and 

<details><summary>void onData(AsyncData<T> value, {bool seamless = false})</summary>
```
final completer = _futureCompleter;
if (completer != null && completer.isCompleted) {
  completer.complete(value.value);
  _futureCompleter = null;
}
```
</details>

But if the completer.isCompleted return true, why trying to complete it again ?
